include/swift/AST/DiagnosticEngine.h

```diff
@@ -799,6 +805,34 @@ namespace swift {
 
   public:
     static const char *diagnosticStringFor(const DiagID id);
+
+    /// If there is no clear .dia file for a diagnostic, put it in the one
+    /// corresponding to the input file name given here.
+    /// In particular, in batch mode when a diagnostic is located in
+    /// a non-primary file, use this affordance to place it in the .dia
+    /// file for the primary that is currently being worked on.
+    void setBufferIndirectlyCausingDiagnosticToInput(
+        StringRef defaultDiagnosticInputFile);
```

Comment:
Hm, why make the DiagnosticEngine do the string lookup? It feels to me like DiagnosticEngine API ought to traffic in SourceLocs.

Reply:
The alternatives that I see are:
1. Expose the `DiagnosticEngine::SourcMgr` instance variable (via a get method?) so that a method in the RAII struct can use it do to the string lookup.
2. Do the lookup in every caller of the RAII class (two, so far).
3. Pass in a SourceManager to the RAII constructor.

Which, if any, are better than what's there now?

Reply:
I was thinking #2 is the best because we probably don't have to do a string lookup in all cases. In particular, SourceFile::getBufferID is available in all cases, right?

(I guess we could make this interface take a buffer ID too, but I don't like that either because it doesn't have a strong type today, and it's also not a type we traffic in in many parts of the compiler. The other advantage of a SourceLoc is that we could some day use it for something more specific.)

Reply:
I tried a couple of alternatives. I quite dislike #2 because we end up with too many lines (for my taste) of duplicated code. Rather, prefer moving the functionality in question to the RAII object, which is sort of a go-between anyway. Also, following your suggestion, the id is obtained without using the filename. So the RAII object gets a `SourceFile`, translates that to a `SourceLoc`, and passes that in to the `DiagnosticEngine`. This layering feels right to me. If you think it best, the RAII definitions and declarations can be moved out of DiagnosticEngine.*. Just tell me where to put them & I'm happy to do it.